### PR TITLE
Add bpf_load (from the kernel) and add userspace loading program and implement most of flowradar

### PR DESCRIPTION
**Pull request by mtbentley**
This requires kernel headers, so for now I'm assuming you have a linux checkout at ../linux.  

This has most of (the server side of) flowradar implemented.

TODO: add "total flow count" somewhere (probably len 1 percpu array?) so we can be sure to swap out maps every time it gets near 8000.

### CODE REVIEW

**Comment by mtbentley** [comment]
The output looks like
```
$ sudo ./dump_maps | jq .                                                  [±add-bpf_load ✓]
{
  "bloomfilter": {
    "0": {
      "0x106": "0x8000000000000",
      "0x166": "0x2000000000000",
      "0x1d6": "0x4000000000000000",
      "0x278": "0x400000000000000",
      "0x2b0": "0x800",
      "0x3c4": "0x8000000000"
    },
    "1": {
      "0xde": "0x40000000000000",
      "0x106": "0x8000000000000",
      "0x14d": "0x4000",
      "0x166": "0x2000000000000",
      "0x1d6": "0x4000000000000000",
      "0x1de": "0x1000",
      "0x278": "0x400000000000000",
      "0x27e": "0x800000000",
      "0x2b0": "0x800",
      "0x30c": "0x20000000000000",
      "0x320": "0x2000000000000000",
      "0x3c4": "0x8000000000"
    }
  },
  "flow_info": {
    "0": {
      "0xee7": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0x53cb": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0x61fa": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0x8a7e": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0xa671": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0xbe73": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c"
    },
    "1": {
      "0xee7": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0x53cb": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0x61fa": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0x8a7e": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0xa671": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0xbe73": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00"
    }
  },
  "eth_proto_count": {
    "0": {
      "0x0800": "0x40f1b"
    },
    "1": {
      "0x0800": "0x1b712",
      "0x0806": "0x3"
    }
  },
  "ip_proto_count": {
    "0": {
      "0x0006": "0x2f00c",
      "0x0001": "0x11f0f"
    },
    "1": {
      "0x0006": "0x14f81",
      "0x0001": "0x6791"
    }
  },
  "sport_count": {
    "0": {
      "0x1389": "0x2f00c"
    },
    "1": {
      "0x1389": "0x14f81"
    }
  },
  "dport_count": {
    "0": {
      "0x963a": "0x2f00c"
    },
    "1": {
      "0x9638": "0x2",
      "0x963a": "0x14f7f"
    }
  },
  "sip_count": {
    "0": {
      "0x400000a": "0x2f00c"
    },
    "1": {
      "0x400000a": "0x14f81"
    }
  },
  "dip_count": {
    "0": {
      "0x100000a": "0x2f00c"
    },
    "1": {
      "0x100000a": "0x14f81"
    }
  },
  "host_info": {
    "0": {
      "0x0": "host=0x2775"
    },
    "1": {
      "0x0": "host=0x2775"
    }
  }
}
```

File: Makefile [hunk 1]
```diff
@@ -19,9 +43,13 @@ clean: unload
 	sudo rm /var/run/netns/h1 || true
 	rm xdp-flowradar.o || true
 	rm test-hash || true
+	rm xdp-flowradar || true
```
Comment (dbgrigsby) [inline Makefile]: Curious, why `rm X || true` insead of `rm -f X`?

File: Makefile [hunk 2]
```diff
@@ -1,12 +1,36 @@
+# NOTE: this requires a kernel with tools/lib/bpf bulit (for libbpf.so)
+LINUX_SOURCE=../linux
+# ^^ is there a better way to do this??
```
Comment (dbgrigsby) [inline Makefile:5]: I don't see a way to do this that is simpler for a dev than the current without being slower (downloading relevant bits in the makefile) or more messy (requiring a specific subfolder of the kernel)

File: bpf_load.c [hunk 3]
```diff
@@ -0,0 +1,669 @@
+// SPDX-License-Identifier: GPL-2.0
```
Comment (dbgrigsby) [inline bpf_load.c:1]: Should add sourcing: https://github.com/MarvellEmbeddedProcessors/mainline-public/blob/master/samples/bpf/bpf_load.c

I spent a while reading this before realizing some of the comments were likely not by you

File: xdp-flowradar_kern.c [hunk 4]
```diff
@@ -93,31 +128,26 @@ struct five_tuple {
...
-	hash = hash * a + daddr1;
-	a = a * b;
-	hash = hash * a + daddr2;
-	a = a * b;
-	hash = hash * a + ft->sport;
-	a = a * b;
-	hash = hash * a + ft->dport;
-	a = a * b;
-	hash = hash * a + ft->proto;
+#pragma unroll
```
Comment (dbgrigsby) [inline xdp-flowradar_kern.c:138]: Maybe this is a common C practice, but why are these loops specifically unrolled and not others?

**Review by dbgrigsby — APPROVED** [review]

File: bpf_load.c (hunk 3 above)
Comment (mtbentley) [reply bpf_load.c:1]: It's actually from `linux/samples/bpf/bpf_load.[ch]`

File: xdp-flowradar_kern.c (hunk 4 above)
Comment (mtbentley) [reply xdp-flowradar_kern.c:138]: The `_kern.c` file gets compiled into the bpf program, which can't have loops (the verifier will reject any jump to a previous instruction).

Normally `#pragma unroll` would be rather unusual, but in this situation we need it.

**Pull request merged**